sourmash_lib/signature_json.py

```diff
@@ -0,0 +1,350 @@
+from __future__ import print_function
+
+import sys
+import warnings
+if sys.version_info[0] < 3:
+    warnings.warn("The module 'signature_json' was written for Python 3 and you Python version is older.")
+
+import sourmash_lib
+from sourmash_lib.signature import FakeHLL, SourmashSignature, SIGNATURE_VERSION
+
+import io
+import json
+import ijson
+
+def _json_next_atomic_array(iterable, prefix_item = 'item', ijson = ijson):
+    """
+    - iterable: iterator as returned by ijson.parse
+    - prefix_item: prefix found for items in the JSON array
+    - ijson: ijson backend
+    """
+    l = list()
+    prefix, event, value = next(iterable)
+    while event != 'start_array':
+        prefix, event, value = next(iterable)
+    prefix, event, value = next(iterable)
+    while event != 'end_array':
+        assert prefix == prefix_item
+        l.append(value)
+        prefix, event, value = next(iterable)
+    return tuple(l)
+
+def test__json_next_atomic_array():
+    t = (2,3,4,5,6)
+    s = json.dumps(t)
+    it = ijson.parse(io.StringIO(s))
+    a = _json_next_atomic_array(it)
+    assert len(t) == len(a)
+    assert all(x == y for x,y in zip(t, a))
+
+def _json_next_signature(iterable,
+                         email = None,
+                         name = None,
+                         filename = None,
+                         ignore_md5sum=False,
+                         prefix_item='mins.item',
+                         ijson = ijson):
+    """Helper function to unpack and check one signature block only.
+    - iterable: an iterable such the one returned by ijson.parse()
+    - email: 
+    - name:
+    - filename:
+    - ignore_md5sum:
+    - prefix_item: required when parsing nested JSON structures
+    - ijson: ijson backend to use.
+    """
+    d = dict()
+    prefix, event, value = next(iterable)
+    if event == 'start_map':
+        prefix, event, value = next(iterable)
+    while event != 'end_map':
+        assert event == 'map_key'
+        key = value
+        if key == 'mins':
+            value = _json_next_atomic_array(iterable,
+                                            prefix_item=prefix_item, ijson=ijson)
+        else:
+            prefix, event, value = next(iterable)
+        d[key] = value
+        prefix, event, value = next(iterable)
+        
+    ksize = d['ksize']
+    mins = d['mins']
+    n = d['num']
+    e = sourmash_lib.Estimators(ksize=ksize, n=n)
+    for m in mins:
+        e.mh.add_hash(m)
+    if 'cardinality' in d:
+        e.hll = FakeHLL(d['cardinality'])
+
+    sig = SourmashSignature(email, e)
+
+    if not ignore_md5sum:
+        md5sum = d['md5sum']
+        if md5sum != sig.md5sum():
+            raise Exception('error loading - md5 of estimator does not match')
+
+    if name:
+        sig.d['name'] = name
+    if filename:
+        sig.d['filename'] = filename
+
+    return sig
+
+# integration test more than a unit test...
+def test__json_next_signature():
```

Comment:
We can move tests to `sourmash_lib/test_signature.py`, but I can also do that later.